scarb/src/ops/package.rs

```diff
@@ -88,10 +96,53 @@ fn before_package(ws: &Workspace<'_>) -> Result<()> {
     Ok(())
 }
 
-#[tracing::instrument(level = "trace", skip(_opts, ws))]
+#[derive(Serialize)]
+struct VcsInfo {
+    sha1: String,
+    path_in_vcs: String,
+}
+
+#[allow(clippy::dbg_macro)]
+fn extract_vcs_info(pkg: &Package, opts: &PackageOpts) -> Result<Option<VcsInfo>> {
+    if let Ok(repo) = gix::discover(pkg.root()) {
+        // There is nothing to parse if repository is empty, so HEAD reference needs to exist.
+        if let Ok(sha1) = repo.rev_parse_single("HEAD") {
+            // `git status -s` output is empty only if there are no changes at all, but always returns success.
+            // `git diff-index --quiet HEAD` returns status code correctly, but doesn't take untracked files into account.
+            let output = Command::new("git")
+                .current_dir(pkg.root())
+                .arg("status")
+                .arg("-s")
+                .output()?;
```

Comment:
This piece of code mixes two stages that the general logic of packaging does:
1. First it tries to **detect** what files are to be included.
2. Second, **only if doing packaging proper**, it computes values of generated files.

This function immediately detects and then computes the value; and then in the `if let Some(vcs_info) = ...` block later on even serializes this file, even in the **first** stage; which is unnecessary.

The boundary between this stages is: 1. is what happens when preparing the recipe, 2. is what happens within the closure passed to `Generated` variant in the recipe file.

Computing the file (i.e. calling `git`) and serializing it should happen in `generated` closure.